homeassistant/components/mqtt/__init__.py

```diff
@@ -245,42 +195,28 @@ async def _async_config_entry_updated(hass: HomeAssistant, entry: ConfigEntry) -
     await hass.config_entries.async_reload(entry.entry_id)
 
 
-async def async_fetch_config(
-    hass: HomeAssistant, entry: ConfigEntry
-) -> dict[str, Any] | None:
+async def async_fetch_config(hass: HomeAssistant, entry: ConfigEntry) -> dict[str, Any]:
     """Fetch fresh MQTT yaml config from the hass config."""
     mqtt_data = get_mqtt_data(hass)
     hass_config = await conf_util.async_hass_config_yaml(hass)
     mqtt_data.config = PLATFORM_CONFIG_SCHEMA_BASE(hass_config.get(DOMAIN, {}))
+    try:
+        entry_config = CONFIG_SCHEMA_ENTRY(dict(entry.data))
+    except vol.MultipleInvalid as ex:
+        raise ConfigEntryError(
+            f"The MQTT config entry is invalid, please correct it: {ex}"
+        ) from ex
 
-    # Remove unknown keys from config entry data
-    _filter_entry_config(hass, entry)
-
-    # Add missing defaults to migrate older config entries
-    await _async_auto_mend_config(hass, entry, mqtt_data.config or {})
-    # Bail out if broker setting is missing
-    if CONF_BROKER not in entry.data:
-        _LOGGER.error("MQTT broker is not configured, please configure it")
-        return None
-
-    # If user doesn't have configuration.yaml config, generate default values
-    # for options not in config entry data
-    if (conf := mqtt_data.config) is None:
-        conf = CONFIG_SCHEMA_ENTRY(dict(entry.data))
-
-    # Merge advanced configuration values from configuration.yaml
-    conf = _merge_extended_config(entry, conf)
-    return conf
+    # Merge yaml config for manual MQTT items with entry data
+    return {**mqtt_data.config, **entry_config}
```

Comment:
I think we can use more modern syntax here:
```suggestion
    return mqtt_data.config | entry_config
```

Reply:
The whole merging is not needed anymore. `entry.data`  is only used for the broker and MQTT client config. YAML is only use for manual configured items.